### ISSUE TITLE
Performance Tuning

### ISSUE THREAD

**Issue by akgunter** [akgunter/crt-royale-reshade#18]
The shader's math is in a place that I'm much more comfortable with. Now it needs performance tuning.

I tested at 2k w/ 3-pixel scanlines and spherical geometry and got the following performance contributions:
- 2ms (15%, 2 passes) on base overhead, crop, and uncrop
- 0.8ms (6%, 1 pass) on interlacing (increases to 3.7ms at 30-pixel scanlines)
- 1.5ms (11.3%, 1 pass) on electron beams
- 0.9ms (6.7%, 1 pass) on beam convergence
- 1.5ms (11.3%, 4 passes) on bloom and blur precalculation
- 1.3ms (9.8%, 2 passes) on deinterlacing and freeze-framing
- 1ms (7.5%, 1 pass + 2 mini passes) on phosphor mask logic
- 0.5ms (3.8%, 1 pass) on the brightpass and color logic
- 2.5ms (18.8%, 2 passes) on bloom application
- 1.3ms (9.8%, 1 pass) on geometry

Then the most expensive passes are:
1. electron beams (11.3%)
2. geometry (9.8%)
3. bloom application (9.4%)
4. interlacing (>=6%)

I don't know how I'll improve performance for these, but I'm sure there's plenty to be done. I could possibly improve performance by combining some passes together, but a lot of the pipelining is necessary.

**Comment by akgunter** [comment]
I've halved compute time for the phosphor mask logic. I might be able to adapt that math to the electron beams and achieve a speedup there too.

**Comment by akgunter** [comment]
Now I've improved the electron beams, approximate-bloom passes, and the interlacing logic. The shader as a whole is roughly 1.5 to 2 times faster now, depending on configs.

Note that some of this is from rewriting the pixelation logic as point sampling instead of averaging. If I re-add averaging as an option, this speedup will be a chunk less impressive when using averaging.

**Comment by akgunter** [comment]
Some notes for myself on old/new performance. Each line is a pass, tested in snes9x in the Super Metroid main menu. 6-pixel scanlines, shadow mask, tested on a Ryzen 5600U and integrated graphics.

These timings are not comparable to the timings in my other post because this benchmark had each pass separated out into its own technique, which increases overhead by a chunk.

```
                                                    OCPU    NCPU    OGPU    NGPU
-   CRT_Royale_0_cropPass                           0.16    -       0.92    -
-   CRT_Royale_1_interlacingPass                    0.09    -       1.26    -
+   CRT_Royale_0_beamDistPass                        -      0.15    -       0.70
    CRT_Royale_2_electronBeamPass                   0.08    0.10    1.65    0.90
    CRT_Royale_3_beamConvergencePass                0.07    0.08    1.17    0.77
    CRT_Royale_4_bloomApproxPassVert                0.07    0.08    1.51    0.78
    CRT_Royale_5_bloomApproxPassHoriz               0.07    0.07    0.83    0.64
    CRT_Royale_6_blurVertialPass                    0.07    0.07    0.63    0.62
    CRT_Royale_7_blurHorizontalPass                 0.06    0.07    0.63    0.78
    CRT_Royale_8_deinterlacePass                    0.07    0.07    0.86    0.77
    CRT_Royale_9_freezeFramePass                    0.06    0.06    0.85    0.61
-   CRT_Royale_10_phosphorMaskResizeVerticalPass    0.07    -       0.72    -
-   CRT_Royale_11_phosphorMaskResizeHorizontalPass  0.06    -       0.85    -
+   CRT_Royale_9_generatePhosphorMask               -       0.06    -       1.04
    CRT_Royale_12_phosphorMaskPass                  0.07    0.06    0.97    0.94
        alias: CRT_Royale_10_applyPhosphorMask			                    
    CRT_Royale_13_brightpassPass                    0.06    0.06    0.96    1.09
    CRT_Royale_14_bloomVerticalPass                 0.06    0.06    1.01    1.09
    CRT_Royale_15_bloomHorizontalPass               0.06    0.06    1.55    1.36
    CRT_Royale_16_geometryPass                      0.05    0.06    1.05    0.95
    CRT_Royale_17_uncropPass                        0.06    0.06    1.16    1.03
```

Collapsing the passes down to one single technique resulted in 12.5ms for the old version and 5.3ms for the new version. This is a 57% speedup on this test.

The brightpass and bloom passes are the next ones I should look at. I also suspect I could improve the uncrop pass by a tiny bit.

**Comment by akgunter** [comment]
I'm gonna say this is good enough for now. I'll add more performance stuff another time.

**Comment by akgunter** [comment]
Released in v2.0.0
